server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -81,7 +81,7 @@ public boolean transferSnapshot(TransferSnapshot transferSnapshot, TranslogTrans
                 translogTransferListener.onUploadComplete(transferSnapshot);
                 return true;
             }
-            final CountDownLatch latch = new CountDownLatch(toUpload.size());
+            final CountDownLatch latch = new CountDownLatch(toUpload.size() + 1);
```

Comment:
I believe we should drop latches and use the `CompletableFuture` instead, the data / metadata transfers could be combined easily (possibly, with support of cancellation at some point).